### PR TITLE
WIP: Fix `-ArgumentList` processing for `Start-Process` for whitespace and embedded quotes

### LINKED ISSUE #5576: Start-Process doesn't pass arguments with embedded whitespace and double quotes correctly
<sup>Updated later to include the problem with embedded double quotes.</sup>

Steps to reproduce
------------------

Embedded whitespace:

```powershell
'"Hi!"' > './t 1.ps1'; Start-Process -Wait -NoNewWindow pwsh -ArgumentList '-noprofile', '-file', './t 1.ps1'
```

Embedded double quotes:

```powershell
Start-Process -Wait -NoNewWindow pwsh -ArgumentList '-noprofile', '-command', '"Hi!"'
```

Expected behavior
-----------------

In both cases:

```none
Hi!
```

That is, script file `./t 1.ps1` should execute, and double-quoted string literal `"Hi!"` should print.

Actual behavior
---------------

Embedded whitespace:

Invocation fails, because the `./t 1.ps1` is passed as _two_ arguments:

```none
The argument './t' is not recognized as the name of a script file. Check the spelling of the name, or if a path was included, verify that the path is
correct and try again.
```

The only way to make this currently work is to _embed_ (potentially escaped) _double quotes_: `'"./t 1.ps1"'` or ``"`"./t 1.ps1`""``; e.g.:

```powershell
Start-Process -Wait -NoNewWindow pwsh -ArgumentList '-noprofile', '-file', "`"./t 1.ps1`""
```

_Update_: Overall, the **best workaround** is to pass a _single_ string containing _all_ arguments to `-ArgumentList` and use embedded quoting and escaping as necessary:

```powershell
Start-Process -Wait -NoNewWindow pwsh -ArgumentList '-noprofile -file "./t 1.ps1"'
```

Embedded double quotes:

The embedded double quotes are unexpectedly _removed_.

The only way to make this currently work is to `\`-escape the embedded double quotes: 

```powershell
Start-Process -Wait -NoNewWindow pwsh -ArgumentList '-noprofile', '-command', '\"Hi!\"'
```

_Update_: Again, the best workaround is to use a _single_ string:

```powershell
Start-Process -Wait -NoNewWindow pwsh -ArgumentList '-noprofile -command \"Hi!\"'
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-rc (v6.0.0-rc) on Microsoft Windows 7 Enterprise  (64-bit; v6.1.7601)
Windows PowerShell v5.1.14409.1012 on Microsoft Windows 7 Enterprise  (64-bit; v6.1.7601)
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

`ProcessStartInfo.ArgumentList` can't be used as the current cmdlet supports: `start-process ping -argumentlist "-n 2 localhost"` which would then treat those 3 arguments as 1 and fail.  

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/5576

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 1]
```diff
@@ -2310,10 +2314,11 @@ private static StringBuilder BuildCommandLine(string executableFileName, string
...
             }
 
-            if (!string.IsNullOrEmpty(arguments))
+            foreach (var arg in arguments)
             {
-                builder.Append(' ');
-                builder.Append(arguments);
+                builder.Append(" \"");
+                builder.Append(arg.Replace("\"","\\\""));
+                builder.Append('"');
```
Comment (felixfbecker) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: Why is any such escaping still needed, vs just letting dotnet handle `StartInfo.ArgumentList`?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 2]
```diff
@@ -4,6 +4,7 @@
 using System;
```
Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: @SteveL-MSFT, your last commit had 4 failures in `PowerShell-CI-macos`
Start-Process.Should handle stdout redirection without error
```powershell
Expected the actual value to be greater than 0, but got 0.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1: line 83
83: 	    $dirEntry.Length | Should -BeGreaterThan 0
```
Start-Process.Should handle arguments with whitespace
```powershell
No such file or directory
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1: line 150
150:         Start-Process -Wait -NoNewWindow testexe.exe -ArgumentList '-echoargs', 'a b', 'c' -RedirectStandardOutput $testdrive/out.txt

```
Start-Process.Should handle arguments with embedded quotes
```powershell
No such file or directory
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1: line 158
158:         Start-Process -Wait -NoNewWindow testexe.exe -ArgumentList '-echoargs', '"a"', 'c' -RedirectStandardOutput $testdrive/out.txt

```
XmlCommand DRT basic functionality Tests.Import-Clixml should work with XML serialization from pwsh.exe
```powershell
Data at the root level is invalid. Line 1, position 1.
at <ScriptBlock>, /Users/runner/runners/2.160.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 110
110: 		$out = Import-Clixml -Path $testfile

```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: @SteveL-MSFT, your last commit had 3 failures in `PowerShell-CI-windows`
WindowStyle argument.-WindowStyle Minimized should work on Windows
```powershell
Expected exactly 'Minimized', but got Hidden.
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 890
890:             $showCmd | Should -BeExactly $WindowStyle
```
WindowStyle argument.-WindowStyle Maximized should work on Windows
```powershell
Expected exactly 'Maximized', but got Hidden.
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 890
890:             $showCmd | Should -BeExactly $WindowStyle
```
XmlCommand DRT basic functionality Tests.Import-Clixml should work with XML serialization from pwsh.exe
```powershell
Data at the root level is invalid. Line 1, position 1.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\XMLCommand.Tests.ps1: line 110
110: 		$out = Import-Clixml -Path $testfile

```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: @SteveL-MSFT, your last commit had 4 failures in `PowerShell-CI-linux`
Start-Process.Should handle stdout redirection without error
```powershell
Expected the actual value to be greater than 0, but got 0.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1: line 83
83: 	    $dirEntry.Length | Should -BeGreaterThan 0
```
Start-Process.Should handle arguments with whitespace
```powershell
No such file or directory
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1: line 150
150:         Start-Process -Wait -NoNewWindow testexe.exe -ArgumentList '-echoargs', 'a b', 'c' -RedirectStandardOutput $testdrive/out.txt

```
Start-Process.Should handle arguments with embedded quotes
```powershell
No such file or directory
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1: line 158
158:         Start-Process -Wait -NoNewWindow testexe.exe -ArgumentList '-echoargs', '"a"', 'c' -RedirectStandardOutput $testdrive/out.txt

```
XmlCommand DRT basic functionality Tests.Import-Clixml should work with XML serialization from pwsh.exe
```powershell
Data at the root level is invalid. Line 1, position 1.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 110
110: 		$out = Import-Clixml -Path $testfile

```

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: It is handled by ArgumentList for Linux, macOS, but Windows uses the Win32 API CreateProcessWithLogon().  Not sure of the history behind this, but preserving the current behavior.

**Comment by SteveL-MSFT** [comment]
Unfortunately, upon closer inspection, I don't see a way to solve the problem without introducing a big breaking change.  Current behavior allows:

```powershell
start-process ping -wait -nonewwindow -argumentlist "-n 2", "localhost"
```

Where all elements of `-ArgumentList` is simply joined into a single string forming: `ping -n 2 localhost`.

We can't start treating them as individual arguments since there is no way to determine what the user intended and arbitrarily add quotes or separating them into individual args.

**Comment by felixfbecker** [comment]
Maybe we could get a new parameter `-RawArgumentList` that is _always_ passed as-is to dotnet `ArgumentList`?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 3]
```diff
@@ -2310,10 +2314,11 @@ private static StringBuilder BuildCommandLine(string executableFileName, string
                 builder.Append('"');
             }
 
-            if (!string.IsNullOrEmpty(arguments))
+            foreach (var arg in arguments)
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: Please add a comment here why escaping is needed here (i.e., this is Windows only, going through Win32 API...).

**Review by PaulHigin — CHANGES_REQUESTED** [review]

**Comment by SteveL-MSFT** [comment]
@felixfbecker let's continue discussion in the issue

**Pull request closed without merging**